### PR TITLE
Player card

### PR DESCRIPTION
**Pull request by philals**
@pjho  Messages in commits.

### CODE REVIEW

**Comment by philals** [comment]
Most of it is what we discussed today

File: index.html [hunk 1]
```diff
@@ -1,9 +1,20 @@
 <!doctype html>
 <html class="no-js" lang="">
 <head>
+<script src="https://ajax.googleapis.com/ajax/libs/jquery/2.1.3/jquery.min.js"></script>
```
Comment (pjho) [inline index.html]: You definitely don't need JS for the grid. I've pushed up [where I got to before I read you were keen to do the player card](https://github.com/pjho/Elo-Score-Board/tree/player-card-eg). That branch has the almost full version of bootstrap css included in the main style sheet & no js & the grid is working fine.

While loading static assets from a public cdn has the advantage that if a user has visited a different site with the same asset previously it may be cached in their browser, we also want to have as few http requests as possible. So there is always some trade off in FE optimisation. In this instance I would bundle the css into our own single stylesheet. jQuery I would probably load from cdn but would reference it in the footer (not that we display anything before scripts load anyway but still a good practice). Bootstrap JS I would bundle. However I'm still adamant we don't need the JS so would prefer to remove this block.

Also our users are almost 100% returning users, so we will have the benefit of cached assets in any case... 

Good catch on the Xero, thought I had removed all refs to the company

File: src/components/PlayerDash.jsx [hunk 2]
```diff
@@ -30,11 +31,14 @@ module.exports = React.createClass({
     // console.log(this.state.players);
     const player = _.find(this.state.players, p => p.id == this.props.params.playerId);
     return (
-      <div className="Player">
-        { player
-          ? JSON.stringify(player)
-          : "not found"
-        }
+      <div className="container-fluid">
```
Comment (pjho) [inline src/components/PlayerDash.jsx]: `<div className="container-fluid">`
The container-fluid class is on the root element in index.html so don't need to duplicate here.

File: src/components/playerdash/player-card.jsx [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    return (
+        <div className="panel panel-default">
+          <table className="table">
+            <thead>
+              <tr>
+                <th><img src={image} className="img-circle img-thumbnail" />{name} - {league} League</th>
+              </tr>
+            </thead>
+            <tbody>
+              <tr>
```
Comment (pjho) [inline src/components/playerdash/player-card.jsx]: These could be good to extract to a sub-component as there's lots of repetition.

``` jsx
<tr>
  <td>Best Rating</td>
  <td>{topScore}</td>
</tr>
```

could become 

``` jsx
<KeyValueRow key="Rating" value={score} />
```

File: src/utils/utilities.js [hunk 4]
```diff
@@ -0,0 +1,10 @@
+module.exports = {
```
Comment (pjho) [inline src/utils/utilities.js]: I like the abstraction, happy with it as it sits but if I was to give feedback...  I would expect a function named `percent` to just return the percent - or NaN if an input was incorrect. So this could be 2 functions, `percentOfPlayerWins` which calls `percent` to get the percent value and handles the additional logic.

also... I know it's my code originally, but as this sits we can make this more succinct..

``` js
let total = wins + losses;
return total > 9 ? (Math.round(wins / (total) * 100) || 0) : '-';
```

:smiley:

**Comment by pjho** [comment]
Cool man, looks good. I've put a bunch of comments in, hopefully some reasonable feedback. Um we really shouldn't need JS for the bootstrap grid.. happy to sit down with you at work and see if we can work out what's not working without it..

File: index.html (hunk 1 above)
Comment (philals) [inline index.html]: Cool. I'll double check the JS thing. I 100% behind removing what is not needed. There might have been some other issue with my bad-Bootstrap code.

File: src/components/PlayerDash.jsx (hunk 2 above)
Comment (philals) [inline src/components/PlayerDash.jsx]: Opps lol

**Comment by philals** [comment]
@pjho Cheers for the tips the other day.

I think I covered most of it. Also learnt some good stuff from your example.

File: src/utils/utilities.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+var percent = function(wins, losses){
```
Comment (pjho) [inline src/utils/utilities.js]: With percent  I was thinking its is a universal concept so I would keep all game/app/context specific logic out of it and just have a pure % function. We don't need to handle NaN as JS will do that for us.

``` js
const percent = function(portion, total) {
  return Math.round(portion / total * 100);
}

percent(10, 100) // 10
percent(10,'100') // 10 - '100' typecast to int
percent(10, 20.323) // 49
percent('a string', 10) // NaN - JS handles NaN for us
percent(false, 100) // 0 - false typcast to 0
percent(100, false) //Infinity - false typecast to 0
percent() //NaN
```

All our logic is better in percentOfPlayerWins. I've put a few checks in which covers most 'possible' results... negative values make it through incorrectly & false is typecast to 0. In this instance I'm not too worried about data validation or throwing a meaningful error as our data is all machine driven so there's not really much chance of getting badly formatted data. The only value we do need to account for is `undefined` which could be passed to percent() and will result in NaN which is handled by !!winPercent which casts as false. The isFinite check is probably overkill but doesn't hurt either.

``` js
const percentOfPlayerWins = function(wins, losses) {
  let total = wins + losses;
  let winPercent = percent(wins, total); // returns number, NaN or Infinity/-Infinity
  return total > 9 && !!winPercent && isFinite(winPercent) ? winPercent + '%' : '-'; // Ensures we have a sensible result
}

percentOfPlayerWins(20, 30); // "40%"
percentOfPlayerWins(undefined, 100); // "-"
percentOfPlayerWins("asdasd", 100); // "-"
percentOfPlayerWins(false, 100); // "-"
percentOfPlayerWins(Infinity,100); // "-"
percentOfPlayerWins(100, false); // "100%" - kindof an issue
percentOfPlayerWins(100, 0); // "100%"
percentOfPlayerWins(100, -0); // "100%"  - kindof an issue
percentOfPlayerWins(-100, -0); // "-"  - kindof an issue
percentOfPlayerWins(-100, 100); // "-"
percentOfPlayerWins(-100, 10); // "-"
percentOfPlayerWins(); // "-"
```

File: src/components/playerdash/player-card.jsx [hunk 6]
```diff
@@ -0,0 +1,36 @@
+import React from 'react';
+import utils from '../../utils/utilities.js';
+import CardLine from './player-card-table-line';
```
Comment (pjho) [inline src/components/playerdash/player-card.jsx]: The only issue I have with `CardLine` as a name is if you want to use the component again later in a different context it may not make sense. I think it's all good for now. If it gets reused at a later date we can refactor and name it something universal and move it from the `playerdash` dir. It could probably be extended to be `row` fairly easily and be passed an array of fields so it could be any length. No need to worry about doing that now though.. :smile:

**Comment by philals** [comment]
@pjho Wicked. All your points make good sense. I have changed most things.

I have renamed the table-line to make it a bit easier if it needs to be reused.

**Comment by pjho** [comment]
:+1:  Cool man, looks good! Looking forward to seeing the graphs :smile:
